### ISSUE TITLE
allow fetching individual object's public url

### ISSUE THREAD

**Issue by 5paceToast** [minio/mc#3027]
## Expected behavior

It is possible to fetch an object's public url without side-effects.

## Actual behavior

The only current way (from what I can tell) to get an individual object's public url is using the following command:
`mc share download --json --expire=1s $OBJECT | jq -r .url`
This is undesirable because it creates a presigned url (even if it expires almost immediately).

The alternatives:
1. `mc policy -r links $OBJECT` -> this does not produce any output against an individual object, with or without `-r`, it appears to only enumerate whole buckets/prefixes.
2. `mc ls --json $OBJECT` -> does not have such a field
3. `mc find $OBJECT --print {url}` -> equivalent to share download, but with a long expiry

## Steps to reproduce the behavior

Go through the documentation and find no alternatives.

## mc --version
- `mcli version RELEASE.2019-12-17T23-26-28Z`

## System information
Arch Linux amd64

## Possible Solutions
a) allow `mc policy links $OBJECT` to run against individual objects
b) add public url as an information point under `mc ls --json`, if the object in question is an object and not a directory (and, potentially, has a download policy)
c) add a {puburl} property to `mc find`'s print
d) have some way to print a host's prefix (e.g https://play.min.io:9000), for both config-based hosts and environment-configured hosts

## Alternate Scenario
It's possible that this functionality already exists, but is either not documented, or documented in an unusual place, in which case this bug should be about improving the documentation to make it more obvious.

**Comment by kannappanr** [comment]
@5paceToast If you take the case of MinIO, it always supports path style.

So, if your endpoint let's say is `https://play.min.io:9000` and the bucket name is `bucket` , then you can do something like 

```
$ mc ls play/bucket --json | jq -r .key
issue
```

The URL will be `https://play.min.io:9000/bucket/issue`

Not sure if this answers your question.

**Comment by 5paceToast** [comment]
> The URL will be `https://play.min.io:9000/bucket/issue`
> 
> Not sure if this answers your question.

No, it does not.
I cannot get `https://play.min.io:9000` programmatically, and this is for a system under which people are supposed to use their own configuration (whether config file or env).
I know you can just stick the object path after the host/etc prefix, the issue is getting the host/etc prefix to work with programmatically (ideally *without* creating a presigned url every time).

An object's public path is always `PROTOCOL :// HOST [: PORT] / BUCKET / PATH`.
I'm interested in being able to get either the entirety of that, or at least a static `PROTOCOL :// HOST [: PORT]`, and it should work whether the configuration was done via `mc config` or `MC_HOST_target` env.
With `MC_HOST_*` you could theoretically use that value, but that would contain credentials, which is less than ideal.

**Comment by kannappanr** [comment]
For aliases on config files, you can do `$ mc config host list play --json | jq -r .URL 
https://play.minio.io:9000`

**Comment by 5paceToast** [comment]
> For aliases on config files

This would have to work on both config files *and* `MC_HOST_target` environment variables the same way (as it does with `mc share download`).
```sh
export MC_HOST_testname="https://Q3AM3UQ867SPQQA43P2F:zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG@play.min.io"
mc config host list testname
# -> No such alias found
```
As such this solution is insufficient.

I'm not necessarily sure what the hold-up is though.
The code to access this information definitely already exists as a part of `mc share download`.
Surely it might be possible to transplant it elsewhere (whether into `ls`, `find`, `config`, `policy links`, or whatever).

**Comment by harshavardhana** [comment]
@5paceToast from what it looks like you want to write your own code, that is a better approach here than using `mc`. `mc` might not be able to satisfy all your requirements.

**Comment by 5paceToast** [comment]
> @5paceToast from what it looks like you want to write your own code
this is for a plugin to a shell script (anypaste.xyz), so saying "install binary X to avoid needing binary Y" isn't really an option

but, again, I very much could try to just plug it into the right part
but I'm not sure which location for the insertion would be accepted, so instead I'm opening it up for the discussion here
is there some deep philosophical reason as for why this might be a fundamental problem?

I mean, consider `policy links` - just making it possible to run against a single object should be fine.
Or consider `ls`, a simple check to see if the policy of the object is "download", in which case providing a url.
Or consider `find`, just having a {puburl} property, etc.

If there is a consensus as to *where* it *should* go, I might tackle adding it in.
But this is a fairly simple use-case, and I really don't see why it would need to be so difficult to do it.

Also as a reminder, there *is* a way to get this information - generate a presigned "download" link that expires in one second. That just returns a public url property on top of the presigned url, and that can be used freely.

In short.
Yes, currently mc does not satisfy the requirements (unless you're willing to constantly generate one-second-long presigned access tokens).
This is literally what this bug is about.

**Comment by 5paceToast** [comment]
For instance, I took a shot at patching the behavior into `mc ls`.
It's a two-line patch (not tested, it's 10pm here, will test tomorrow if no counter-indication shows up here).
Looks like this against master:
```diff
diff --git a/cmd/ls.go b/cmd/ls.go
index bbf2ddc8..d480d063 100644
--- a/cmd/ls.go
+++ b/cmd/ls.go
@@ -42,6 +42,7 @@ type contentMessage struct {
 	Size     int64     `json:"size"`
 	Key      string    `json:"key"`
 	ETag     string    `json:"etag"`
+	URL      string    `json:"url"`
 }
 
 // String colorized string message.
@@ -80,6 +81,7 @@ func parseContent(c *clientContent) contentMessage {
 	}()
 
 	content.Size = c.Size
+	content.URL = c.URL.String()
 	md5sum := strings.TrimPrefix(c.ETag, "\"")
 	md5sum = strings.TrimSuffix(md5sum, "\"")
 	content.ETag = md5sum
```

**Comment by 5paceToast** [comment]
As a further update, this didn't work because the value of `c.URL` was changed before passing it to the function around lines 140 and onward.
To keep backwards compatibility, I marked the field as optional in json and moved the calculation to the worker function.
```diff
diff --git a/cmd/ls.go b/cmd/ls.go
index bbf2ddc8..a2b5a8ae 100644
--- a/cmd/ls.go
+++ b/cmd/ls.go
@@ -42,6 +42,7 @@ type contentMessage struct {
 	Size     int64     `json:"size"`
 	Key      string    `json:"key"`
 	ETag     string    `json:"etag"`
+	URL      string    `json:"url,omitempty"`
 }
 
 // String colorized string message.
@@ -148,6 +149,9 @@ func doList(clnt Client, isRecursive, isIncomplete bool) error {
 		contentURL = strings.TrimPrefix(contentURL, prefixPath)
 		content.URL.Path = contentURL
 		parsedContent := parseContent(content)
+		// URL is empty by default
+		// Set it to either relative dir (host) or public url (remote)
+		parsedContent.URL = clnt.GetURL().String()
 		// Print colorized or jsonized content info.
 		printMsg(parsedContent)
 	}
```
This is the result, and it functions as expected.
./cmd/... tests also pass just fine (should be all that needs to pass given that that's the only changed package).
